### PR TITLE
Harry & Chris: Model Implementation

### PR DESCRIPTION
**Pull request by hburn7**
Prerequisites:
- [x] Merge of #4 into master

Milestones to meet before this PR will be open to review:
- [x] Create basic architecture / directory structure for TensorFlow model
- ~~Create custom TensorFlow object recognition model that is able to be trained (under core.model)~~
- ~~[ ] Create a basic testing suite for the model (Does it run? Does it trigger correctly based on program args? Does it produce an accurate result based on training data?)~~

### CODE REVIEW

File: src/core/model/recognizer.py [hunk 1]
```diff
@@ -41,10 +42,13 @@ def draw_bounding_box_on_image(image, y_min, x_min, y_max, x_max, color, font, t
...
+    # the tutorial this came from had the same thing.
+    # I think the tutorial meant to put fig instead of plt on the rest of the method.
+    # hopefully changing it to fig will make it play nice
     fig = plt.figure(figsize=(20, 15))
-    plt.grid(False)
-    plt.imshow(image)
-    plt.show()
+    fig.grid(False)
+    fig.imshow(image)
+    fig.show()
```
Comment (hburn7) [inline src/core/model/recognizer.py]: Thanks for this fix. Surprised I didn't catch that.

**Comment by hburn7** [comment]
Alright, here goes. As of this comment, the following changes have been made.

- Full implementation of FiftyOne for purposes of processing unlabeled image data against a pre-trained COCO dataset.
- Full implementation of image labeling
- Full implementation of image rendering and output.
- Full Dockerization along with documentation on how to run the Docker image.
- Documentation on known caveats of this program (i.e. unable to run on Windows from source)
- Major refactoring to the architecture of the program (removed 4 different CLI args and added capability to run without any args).

**Review by theguy951357 — APPROVED** [review]

**Review by mr-matthew-michael — APPROVED** [review]

**Comment by jaydinandrews** [comment]
Few things noticed when testing:
1. Testing on an old Linux machine, I do not have a GPU, and the `--gpus all` flag throws an error. Not really a big deal since it runs fine on my old Intel i7-3520M in under a minute.
2. The `./out/` directory that is created is owned by root and not modifiable by my user without superuser privileges. This could be due to a docker daemon issue, can anyone else confirm this? 

The output of `docker run` without the `--gpus all` flag:
```
2022-02-28 18:46:12,112 - main - [DEBUG]: Logger initialized. @(main.py:40)
Welcome to YAOR (Yet Another Object Recognizer)!
2022-02-28 18:46:12,113 - main - [INFO]: Welcome to YAOR (Yet Another Object Recognizer)! @(main.py:61)
Loaded with configuration: Config(image_dir=./images/, out_dir=./out/, verbose=True)
2022-02-28 18:46:12,114 - main - [INFO]: Loaded with configuration: Config(image_dir=./images/, out_dir=./out/, verbose=True) @(main.py:62)
Checking TensorFlow for hardware access. (Do you see 'GPU'?)...
2022-02-28 18:46:12,114 - main - [INFO]: Checking TensorFlow for hardware access. (Do you see 'GPU'?)... @(main.py:52)
2022-02-28 18:46:12.114703: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: UNKNOWN ERROR (34)
2022-02-28 18:46:12.114752: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (c35ea81b6630): /proc/driver/nvidia/version does not exist
TensorFlow has access to the following devices:
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]
2022-02-28 18:46:12,114 - main - [INFO]: TensorFlow has access to the following devices:
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')] @(main.py:54)
TensorFlow version: 2.8.0
2022-02-28 18:46:12,114 - main - [INFO]: TensorFlow version: 2.8.0 @(main.py:57)
Begin image labeling.
2022-02-28 18:46:12,115 - main - [INFO]: Begin image labeling. @(main.py:66)
2022-02-28 18:46:12,123 - universal_analytics.requests - [DEBUG]: Request: POST https://www.google-analytics.com/collect; payload=el=0.14.4-NONE&t=event&ec=import&ea=new&v=1&tid=UA-141773487-9&cid=5c42f71b-2b77-4841-b766-2aaf8c917965 @(requests.py:64)
Downloading model from 'http://download.tensorflow.org/models/object_detection/tf2/20200711/centernet_resnet50_v2_512x512_coco17_tpu-8.tar.gz'...
2022-02-28 18:46:12,174 - fiftyone.core.models - [INFO]: Downloading model from 'http://download.tensorflow.org/models/object_detection/tf2/20200711/centernet_resnet50_v2_512x512_coco17_tpu-8.tar.gz'... @(models.py:1787)
2022-02-28 18:46:12,176 - urllib3.connectionpool - [DEBUG]: Starting new HTTP connection (1): download.tensorflow.org:80 @(connectionpool.py:221)
2022-02-28 18:46:12,288 - httpx._client - [DEBUG]: HTTP Request: POST https://www.google-analytics.com/collect "HTTP/1.1 200 OK" @(_client.py:1002)
2022-02-28 18:46:12,505 - urllib3.connectionpool - [DEBUG]: http://download.tensorflow.org:80 "GET /models/object_detection/tf2/20200711/centernet_resnet50_v2_512x512_coco17_tpu-8.tar.gz HTTP/1.1" 200 237977273 @(connectionpool.py:428)
 100%     1.8Gb/1.8Gb [22.2s elapsed, 0s remaining, 84.5Mb/s]      
2022-02-28 18:46:34,708 - eta.core.utils - [INFO]:  100%     1.8Gb/1.8Gb [22.2s elapsed, 0s remaining, 84.5Mb/s]       @(utils.py:1921)
Extracting archive '/root/fiftyone/__models__/centernet_resnet50_v2_512x512_coco17_tpu-8.tar.gz'
2022-02-28 18:46:34,709 - eta.core.models - [INFO]: Extracting archive '/root/fiftyone/__models__/centernet_resnet50_v2_512x512_coco17_tpu-8.tar.gz' @(models.py:1635)
2022-02-28 18:46:36,961 - eta.core.utils - [DEBUG]: Executing 'lspci' @(utils.py:2164)
 100%  2/2 [10.0ms elapsed, 0s remaining, 200.3 samples/s]   
2022-02-28 18:46:37,173 - eta.core.utils - [INFO]:  100%  2/2 [10.0ms elapsed, 0s remaining, 200.3 samples/s]    @(utils.py:1921)
   0%  0/2 [4.4ms elapsed, ? remaining, ? samples/s] 2022-02-28 18:47:00.164032: W tensorflow/core/grappler/optimizers/loop_optimizer.cc:907] Skipping loop optimization for Merge node with control input: StatefulPartitionedCall/Assert/AssertGuard/branch_executed/_336
 100%  2/2 [11.5s elapsed, 0s remaining, 0.2 samples/s]  
2022-02-28 18:47:08,073 - eta.core.utils - [INFO]:  100%  2/2 [11.5s elapsed, 0s remaining, 0.2 samples/s]   @(utils.py:1921)
Directory './out/' already exists; outputs will be merged with existing files
2022-02-28 18:47:08,075 - fiftyone.core.collections - [WARNING]: Directory './out/' already exists; outputs will be merged with existing files @(collections.py:5920)
   0%  0/2 [7.5ms elapsed, ? remaining, ? samples/s] 2022-02-28 18:47:08,124 - eta.core.annotations - [DEBUG]: Rendering 25 objects @(annotations.py:858)
  50%  1/2 [635.1ms elapsed, 635.1ms remaining, 1.6 samples/s] 2022-02-28 18:47:08,765 - eta.core.annotations - [DEBUG]: Rendering 10 objects @(annotations.py:858)
 100%  2/2 [1.1s elapsed, 0s remaining, 1.9 samples/s]         
2022-02-28 18:47:09,133 - eta.core.utils - [INFO]:  100%  2/2 [1.1s elapsed, 0s remaining, 1.9 samples/s]          @(utils.py:1921)
```

**Comment by hburn7** [comment]
> Few things noticed when testing:
> 
> 1. Testing on an old Linux machine, I do not have a GPU, and the `--gpus all` flag throws an error. Not really a big deal since it runs fine on my old Intel i7-3520M in under a minute.
> 2. The `./out/` directory that is created is owned by root and not modifiable by my user without superuser privileges. This could be due to a docker daemon issue, can anyone else confirm this?
> 
> The output of `docker run` without the `--gpus all` flag:
> 
> ```
> 2022-02-28 18:46:12,112 - main - [DEBUG]: Logger initialized. @(main.py:40)
> Welcome to YAOR (Yet Another Object Recognizer)!
> 2022-02-28 18:46:12,113 - main - [INFO]: Welcome to YAOR (Yet Another Object Recognizer)! @(main.py:61)
> Loaded with configuration: Config(image_dir=./images/, out_dir=./out/, verbose=True)
> 2022-02-28 18:46:12,114 - main - [INFO]: Loaded with configuration: Config(image_dir=./images/, out_dir=./out/, verbose=True) @(main.py:62)
> Checking TensorFlow for hardware access. (Do you see 'GPU'?)...
> 2022-02-28 18:46:12,114 - main - [INFO]: Checking TensorFlow for hardware access. (Do you see 'GPU'?)... @(main.py:52)
> 2022-02-28 18:46:12.114703: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: UNKNOWN ERROR (34)
> 2022-02-28 18:46:12.114752: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (c35ea81b6630): /proc/driver/nvidia/version does not exist
> TensorFlow has access to the following devices:
> [PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]
> 2022-02-28 18:46:12,114 - main - [INFO]: TensorFlow has access to the following devices:
> [PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')] @(main.py:54)
> TensorFlow version: 2.8.0
> 2022-02-28 18:46:12,114 - main - [INFO]: TensorFlow version: 2.8.0 @(main.py:57)
> Begin image labeling.
> 2022-02-28 18:46:12,115 - main - [INFO]: Begin image labeling. @(main.py:66)
> 2022-02-28 18:46:12,123 - universal_analytics.requests - [DEBUG]: Request: POST https://www.google-analytics.com/collect; payload=el=0.14.4-NONE&t=event&ec=import&ea=new&v=1&tid=UA-141773487-9&cid=5c42f71b-2b77-4841-b766-2aaf8c917965 @(requests.py:64)
> Downloading model from 'http://download.tensorflow.org/models/object_detection/tf2/20200711/centernet_resnet50_v2_512x512_coco17_tpu-8.tar.gz'...
> 2022-02-28 18:46:12,174 - fiftyone.core.models - [INFO]: Downloading model from 'http://download.tensorflow.org/models/object_detection/tf2/20200711/centernet_resnet50_v2_512x512_coco17_tpu-8.tar.gz'... @(models.py:1787)
> 2022-02-28 18:46:12,176 - urllib3.connectionpool - [DEBUG]: Starting new HTTP connection (1): download.tensorflow.org:80 @(connectionpool.py:221)
> 2022-02-28 18:46:12,288 - httpx._client - [DEBUG]: HTTP Request: POST https://www.google-analytics.com/collect "HTTP/1.1 200 OK" @(_client.py:1002)
> 2022-02-28 18:46:12,505 - urllib3.connectionpool - [DEBUG]: http://download.tensorflow.org:80 "GET /models/object_detection/tf2/20200711/centernet_resnet50_v2_512x512_coco17_tpu-8.tar.gz HTTP/1.1" 200 237977273 @(connectionpool.py:428)
>  100%     1.8Gb/1.8Gb [22.2s elapsed, 0s remaining, 84.5Mb/s]      
> 2022-02-28 18:46:34,708 - eta.core.utils - [INFO]:  100%     1.8Gb/1.8Gb [22.2s elapsed, 0s remaining, 84.5Mb/s]       @(utils.py:1921)
> Extracting archive '/root/fiftyone/__models__/centernet_resnet50_v2_512x512_coco17_tpu-8.tar.gz'
> 2022-02-28 18:46:34,709 - eta.core.models - [INFO]: Extracting archive '/root/fiftyone/__models__/centernet_resnet50_v2_512x512_coco17_tpu-8.tar.gz' @(models.py:1635)
> 2022-02-28 18:46:36,961 - eta.core.utils - [DEBUG]: Executing 'lspci' @(utils.py:2164)
>  100%  2/2 [10.0ms elapsed, 0s remaining, 200.3 samples/s]   
> 2022-02-28 18:46:37,173 - eta.core.utils - [INFO]:  100%  2/2 [10.0ms elapsed, 0s remaining, 200.3 samples/s]    @(utils.py:1921)
>    0%  0/2 [4.4ms elapsed, ? remaining, ? samples/s] 2022-02-28 18:47:00.164032: W tensorflow/core/grappler/optimizers/loop_optimizer.cc:907] Skipping loop optimization for Merge node with control input: StatefulPartitionedCall/Assert/AssertGuard/branch_executed/_336
>  100%  2/2 [11.5s elapsed, 0s remaining, 0.2 samples/s]  
> 2022-02-28 18:47:08,073 - eta.core.utils - [INFO]:  100%  2/2 [11.5s elapsed, 0s remaining, 0.2 samples/s]   @(utils.py:1921)
> Directory './out/' already exists; outputs will be merged with existing files
> 2022-02-28 18:47:08,075 - fiftyone.core.collections - [WARNING]: Directory './out/' already exists; outputs will be merged with existing files @(collections.py:5920)
>    0%  0/2 [7.5ms elapsed, ? remaining, ? samples/s] 2022-02-28 18:47:08,124 - eta.core.annotations - [DEBUG]: Rendering 25 objects @(annotations.py:858)
>   50%  1/2 [635.1ms elapsed, 635.1ms remaining, 1.6 samples/s] 2022-02-28 18:47:08,765 - eta.core.annotations - [DEBUG]: Rendering 10 objects @(annotations.py:858)
>  100%  2/2 [1.1s elapsed, 0s remaining, 1.9 samples/s]         
> 2022-02-28 18:47:09,133 - eta.core.utils - [INFO]:  100%  2/2 [1.1s elapsed, 0s remaining, 1.9 samples/s]          @(utils.py:1921)
> ```

@jaydinandrews Hey Jay, thanks for taking a look and for testing out the code on your machine. At this time I cannot guarantee support for running without a supported NVIDIA graphics card as the docker image relies on tensorflow-gpu. Based on what I see in the console output here, it looks like things ran without issue. My main concern with CPU-only mode is that it produces bad labels. Can you `chown` the `out` directory such that your user has access and verify the output looks okay? Also, I appreciate your note on the folder permissions, that is something I can look into. My guess at a quick solution is creating the folders first through your user account (`mkdir out images`) rather than having Docker do that automatically. That way the folder can be mapped into the container without new permissions being generated. I'd appreciate it if you could try that on another run of the Docker image and see if the folder permission issue persists. Thanks!

**Comment by jaydinandrews** [comment]
@hburn7 Yeah, the labels look alright. Label accuracy is low (in the 40%-80%) ranges however nearly all are correct. The label accuracy is not a problem however as they closely resemble the outputted on your previous testing. 
Making the directory before does work however the issue still arises on a clean image. If it is only happening on my machine, I believe it is just a docker daemon problem since by default it can only be run as root and Docker is its own user with high privileges. Maybe a check for these directories in main might be a good idea? The directories would be created by python as the running user instead of the way docker volumes does it. Just a thought, otherwise, everything looks great.

**Review by jaydinandrews — APPROVED** [review]
I will approve and will leave the merge up to you if we wanna sort out the directory check.

**Comment by hburn7** [comment]
@jaydinandrews I like the idea of the directories being created in main, I'll implement that soon. I do want to change the model as well, I've tested several and found an optimal balance between recognition accuracy and speed (should be more accurate than what is present already). After that is taken care of I'll do a cleanup and merge.

**Comment by hburn7** [comment]
Added video support, updated model, README with more instructions, and created necessary directories on startup.
